src/test/java/org/apache/sling/discovery/oak/JoinerDelayTest.java

```diff
@@ -0,0 +1,299 @@
+        PrivateAccessor.setField(scheduler, "scheduler", qScheduler);
+
+        closer.register(new Closeable() {
+
+            @Override
+            public void close() throws IOException {
+                try {
+                    PrivateAccessor.invoke(qScheduler, "deactivate", new Class[] {BundleContext.class}, new Object[] {ctx});
+                } catch (Throwable e) {
+                    throw new IOException(e);
+                }
+            }
+
+        });
+
+        return scheduler;
+    }
+
+    private Scheduler createDummyScheduler() {
+        return new DummyScheduler(true);
+    }
+
+    @Parameterized.Parameters(name="{0}")
+    public static Collection<SchedulerType> schedulerTypes() {
+        Collection<SchedulerType> result = new ArrayList<>();
+        result.add(SchedulerType.REAL);
+        result.add(SchedulerType.DUMMY);
+        return result;
+    }
+
+    public JoinerDelayTest(SchedulerType schedulerType) {
+        this.schedulerType = schedulerType;
+    }
+
+    @Before
+    public void setup() throws Throwable {
+        closer = Closer.create();
+        switch(schedulerType) {
+        case REAL : {
+            scheduler = createRealScheduler();
+            break;
+        }
+        case DUMMY : {
+            scheduler = createDummyScheduler();
+            break;
+        }
+        default: {
+            fail("unknown schedulerType : " + schedulerType);
+        }
+        }
+        DefaultClusterView cluster = new DefaultClusterView(UUID.randomUUID().toString());
+        view = new DummyTopologyView()
+                .addInstance(UUID.randomUUID().toString(), cluster, true, true)
+                .addInstance(UUID.randomUUID().toString(), cluster, false, false);
+        callbackSemaphore = new Semaphore(0);
+        callback = new Runnable() {
+
+            @Override
+            public void run() {
+                callbackSemaphore.release();
+            }
+
+        };
+    }
+
+    @After
+    public void teardown() throws Throwable {
+        closer.close();
+    }
+
+    @Test
+    public void testDummyCancelSyncCalls() throws Exception {
+        JoinerDelay joinerDelay = new JoinerDelay(-1, scheduler);
+        joinerDelay.cancelSync();
+        joinerDelay.cancelSync();
+    }
+
+    @Test
+    public void testSync_nulValues() throws Exception {
+        JoinerDelay joinerDelay = new JoinerDelay(1, scheduler);
+        try {
+            joinerDelay.sync(null, null);
+            fail("should fail");
+        } catch(Exception e) {
+            // ok
+        }
+        joinerDelay = new JoinerDelay(1, scheduler);
+        try {
+            joinerDelay.sync(view, null);
+            fail("should fail");
+        } catch(Exception e) {
+            // ok
+        }
+        joinerDelay = new JoinerDelay(1, scheduler);
+        // this one is fine as it is caught
+        joinerDelay.sync(null, callback);
```

Comment:
I think the implementation should also throw an `IllegalArgumentException` when `view` is `null`. The contract does not mention `view` can be `null` and also does not specify what should happen in that case.